packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx

```diff
@@ -0,0 +1,90 @@
+import * as React from 'react';
+import { useControllableState } from '@fluentui/react-utilities';
+import type { TimePickerProps, TimePickerState, TimeSelectionData } from './TimePicker.types';
+import { ComboboxProps, useCombobox_unstable, Option } from '@fluentui/react-combobox';
+import {
+  dateToKey,
+  keyToDate,
+  formatTimeString,
+  getDateStartAnchor,
+  getDateEndAnchor,
+  getTimesBetween,
+} from './timeMath';
+
+/**
+ * Create the state required to render TimePicker.
+ *
+ * The returned state can be modified with hooks such as useTimePickerStyles_unstable,
+ * before being passed to renderTimePicker_unstable.
+ *
+ * @param props - props from this instance of TimePicker
+ * @param ref - reference to root HTMLElement of TimePicker
+ */
+export const useTimePicker_unstable = (props: TimePickerProps, ref: React.Ref<HTMLInputElement>): TimePickerState => {
+  const {
+    dateAnchor: dateAnchorInProps,
+    defaultSelectedTime: defaultSelectedTimeInProps,
+    endHour = 24,
+    hour12 = false,
+    increment = 30,
+    onTimeSelect,
+    selectedTime: selectedTimeInProps,
+    showSeconds = false,
+    startHour = 0,
+    ...rest
+  } = props;
+
+  const fallbackDateAnchor = React.useRef(new Date()).current;
+  const dateAnchor = dateAnchorInProps ?? selectedTimeInProps ?? defaultSelectedTimeInProps ?? fallbackDateAnchor;
+
+  const dateStartAnchor = React.useMemo(() => getDateStartAnchor(dateAnchor, startHour), [dateAnchor, startHour]);
+  const dateEndAnchor = React.useMemo(
+    () => getDateEndAnchor(dateAnchor, startHour, endHour),
+    [dateAnchor, endHour, startHour],
+  );
+
+  const options: Date[] = React.useMemo(
+    () => getTimesBetween(dateStartAnchor, dateEndAnchor, increment),
+    [increment, dateStartAnchor, dateEndAnchor],
```

Comment:
I think it might be worth using the result of `dateToKey(dateStartAnchor)` and `dateToKey(dateEndAnchor)` as memo deps, since I don't think users would memoize `new Date()` that often

Reply:
I see. I created a new hook `useStableDateAnchor` - please take a look :)